mlos_core/mlos_core/data_classes.py

```diff
@@ -0,0 +1,350 @@
+        Converts the observation to a suggestion.
+
+        Returns
+        -------
+        Suggestion
+            The suggestion.
+        """
+        return Suggestion(
+            config=self._config,
+            context=self._context,
+            metadata=self._metadata,
+        )
+
+    def __repr__(self) -> str:
+        return (
+            f"Observation(config={self._config}, score={self._score},"
+            " context={self._context}, metadata={self._metadata})"
+        )
+
+    def __eq__(self, other: Any) -> bool:
+        if not isinstance(other, Observation):
+            return False
+
+        if not self._config.equals(other._config):
+            return False
+        if not self._score.equals(other._score):
+            return False
+        if not compare_optional_series(self._context, other._context):
+            return False
+        if not compare_optional_series(self._metadata, other._metadata):
+            return False
+
+        return True
+
+    def __ne__(self, other: Any) -> bool:
+        return not self.__eq__(other)
+
+
+class Observations:
+    """
+    A set of observations of a configuration's score.
+
+    Attributes
+    ----------
+    config : pd.DataFrame
+        Pandas dataframe containing configurations. Column names are the parameter names.
+    score : pd.DataFrame
+        The score metrics observed in a dataframe.
+    context : Optional[pd.Series]
+        The context in which the configuration was evaluated.
+        Not Yet Implemented.
+    metadata: Optional[pd.Series]
+        The metadata in which the configuration was evaluated
+        Not Yet Implemented.
+    """
+
+    def __init__(
+        self,
+        *,
+        config: pd.DataFrame = pd.DataFrame(),
+        score: pd.DataFrame = pd.DataFrame(),
+        context: Optional[pd.DataFrame] = None,
+        metadata: Optional[pd.DataFrame] = None,
+        observations: List[Observation] = [],
+    ):
+
+        if len(observations) > 0:
+            config = pd.concat([obs._config.to_frame().T for obs in observations])
+            score = pd.concat([obs._score.to_frame().T for obs in observations])
+
+            if sum([obs._context is None for obs in observations]) == 0:
+                context = pd.concat(
+                    [obs._context.to_frame().T for obs in observations]  # type: ignore
+                )
+            else:
+                context = None
+            if sum([obs._metadata is None for obs in observations]) == 0:
+                metadata = pd.concat(
+                    [obs._metadata.to_frame().T for obs in observations]  # type: ignore
+                )
+            else:
+                metadata = None
+        assert len(config.index) == len(score.index), "config and score must have the same length"
+        if context is not None:
+            l1 = len(config.index)
+            l2 = len(context.index)
+            assert l1 == l2, "config and context must have the same length"
```

Comment:
skip variable assignment, include the `len()` check in the assert and use a line break for the message?